src/Raven.Client/Documents/Session/AsyncSessionDocumentCounters.cs

```diff
@@ -0,0 +1,170 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="AsyncSessionDocumentCounters.cs" company="Hibernating Rhinos LTD">
+//     Copyright (c) Hibernating Rhinos LTD. All rights reserved.
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Raven.Client.Documents.Operations.Counters;
+using Sparrow.Json;
+
+namespace Raven.Client.Documents.Session
+{
+    public class AsyncSessionDocumentCounters : SessionCountersBase, IAsyncSessionDocumentCounters
+    {
+        public AsyncSessionDocumentCounters(InMemoryDocumentSessionOperations session, string documentId) : base(session, documentId)
+        {
+        }
+
+        public AsyncSessionDocumentCounters(InMemoryDocumentSessionOperations session, object entity) : base(session, entity)
+        {
+        }
+
+        public async Task<long?> GetAsync(string counter, CancellationToken token = default)
+        {
+            long? value = null;
+            if (Session.CountersByDocId.TryGetValue(DocId, out var cache))
+            {
+                if (cache.Values.TryGetValue(counter, out value))
+                    return value;
+            }
+            else
+            {
+                cache.Values = new Dictionary<string, long?>(StringComparer.OrdinalIgnoreCase);
+            }
+
+            if ((Session.DocumentsById.TryGetValue(DocId, out var document) == false && cache.GotAll == false) ||
+                (document != null && document.Metadata.TryGet(Constants.Documents.Metadata.Counters, 
+                    out BlittableJsonReaderArray metadataCounters) &&
+                metadataCounters.BinarySearch(counter, StringComparison.OrdinalIgnoreCase) >= 0))
+                
+            {
+                // we either don't have the document in session and GotAll = false,
+                // or we do and it's metadata contains the counter name
+
+                Session.IncrementRequestCount();
+
+                var details = await Session.DocumentStore.Operations.ForDatabase(Session.DatabaseName)
```

Comment:
I think that we have RequestExecutor and OperationExecutor in session already. Try using:
```
Session.Operations.SendAsync
```

Also remember to put `Session.SessionInfo` as a parameter to `SendAsync`.

Reply:
done